File: Queue/queue.rb
class Queue
  attr_reader :add, :remove, :length
  def initialize
    @storage = {}
    @size    = 0
    @front   = 0
  end

  def add(value)
    @storage[@size] = value
    @size += 1
  end
  def remove
    @front += 1
    return @storage[@front-1]
  end
  def length
    return @size-@front
  end
end

File: Queue/queue.js
// Pseudo-Classical pattern

var Queue = function() {
  this.storage = {};
  this.front = 0;
  this.queueSize = 0;
};

Queue.prototype.enqueue = function(value) {
  this.storage[this.queueSize++] = value;
};
Queue.prototype.dequeue = function() {
  if (this.queueSize - this.front) {
    var temp = this.storage[this.front];
    return (delete this.storage[this.front++]) && temp;
  }
};
Queue.prototype.size = function(value) {
  return this.queueSize - this.front;
};
